File: tutorial/blocks.py
# ============================================================
#  blocks.py  –  Clase Bloque (las opciones de respuesta)
#  ¡Casi todo está listo! Aquí se dibujan los bloques de respuestas
# ============================================================

import pygame
import random
from settings import *


class Bloque:
    def __init__(self, cx, cy, texto, es_correcto):
        self.cx = cx          # centro x
        self.cy = cy          # centro y
        self.ancho = BLOQUE_ANCHO
        self.alto  = BLOQUE_ALTO
        self.texto = texto
        self.es_correcto = es_correcto
        self.activo = True
        self.sacudida = 0
        self.destello = 0

    # ---------- LÓGICA ----------

    def verificar_colision(self, bala):
        if not self.activo or bala is None or not bala.activa:
            return False
        # PISTA: Compara las posiciones de la bala y el bloque
        return (abs(bala.x - self.cx) <= self.ancho / 2 + bala.radio and
                abs(bala.y - self.cy) <= self.alto  / 2 + bala.radio)

    def destruir(self):
        self.activo = False

    def sacudir(self):
        # PISTA: ¿Cuántos frames quieres que dure la animación?
        self.sacudida = 25

    def destellar(self):
        self.destello = 20

    def actualizar(self):
        if self.sacudida > 0:
            self.sacudida -= 1
        if self.destello > 0:
            self.destello -= 1

    # ---------- DIBUJO ----------

    def _ajustar_texto(self, fuente):
        palabras = self.texto.split()
        lineas, linea = [], ""
        for p in palabras:
            prueba = (linea + " " + p).strip()
            if fuente.size(prueba)[0] <= self.ancho - 14:
                linea = prueba
            else:
                if linea:
                    lineas.append(linea)
                linea = p
        if linea:
            lineas.append(linea)
        return lineas

    def dibujar(self, pantalla, fuente):
        if not self.activo:
            return

        # Efecto de sacudida (vibración)
        ox = random.randint(-3, 3) if self.sacudida > 0 else 0
        rx = self.cx - self.ancho // 2 + ox
        ry = self.cy - self.alto  // 2

        rect = pygame.Rect(rx, ry, self.ancho, self.alto)

        # ---------- COMPLETAR: Colores ----------
        # PISTA: ¿Qué color quieres cuando está sacudido? ¿Y normalmente?
        #        - Normal: fondo AZUL, borde NEGRO
        #        - Sacudido: fondo ROJO, borde AMARILLO
        color_fondo = ROJO if self.sacudida > 0 else AZUL
        color_borde = AMARILLO if self.sacudida > 0 else NEGRO

        # Efecto de destello (cuando es正确答案)
        if self.destello > 0:
            if (self.destello // 3) % 2 == 0:
                color_fondo = VERDE
                color_borde = NEGRO
            else:
                color_fondo = (180, 255, 180)
                color_borde = VERDE

        # PISTA: Dibuja el rectángulo con borde de 3px
        pygame.draw.rect(pantalla, color_fondo, rect, border_radius=12)
        pygame.draw.rect(pantalla, color_borde, rect, 3, border_radius=12)

        # Texto del bloque
        lineas = self._ajustar_texto(fuente)
        total_h = len(lineas) * fuente.get_height()
        sy = self.cy - total_h // 2

        for linea in lineas:
            surf = fuente.render(linea, True, BLANCO)
            pantalla.blit(surf, surf.get_rect(centerx=self.cx + ox, y=sy))
            sy += fuente.get_height()


# ============================================================
#  💡 TAREA EXTRA:
#  - Prueba cambiar los colores de los bloques
#  - ¿Qué pasa si cambias el tiempo de 'sacudida' o 'destello'?
# ============================================================

File: web/blocks.js
// ============================================================
//  blocks.js  –  Clase Bloque (opciones de respuesta)
// ============================================================

import { BLOQUE_ANCHO, BLOQUE_ALTO, AZUL, ROJO, VERDE, BLANCO, NEGRO_PURO } from './settings.js';

export class Bloque {
    constructor(cx, cy, texto, esCorrecto) {
        this.cx = cx;
        this.cy = cy;
        this.ancho = BLOQUE_ANCHO;
        this.alto = BLOQUE_ALTO;
        this.texto = texto;
        this.esCorrecto = esCorrecto;
        this.activo = true;
        this.sacudida = 0;
        this.destello = 0;
    }

    verificarColision(bala) {
        if (!this.activo || bala === null || !bala.activa) return false;

        return (
            Math.abs(bala.x - this.cx) <= this.ancho / 2 + bala.radio &&
            Math.abs(bala.y - this.cy) <= this.alto / 2 + bala.radio
        );
    }

    destruir() {
        this.activo = false;
    }

    sacudir() {
        this.sacudida = 25;
    }

    destellar() {
        this.destello = 20;
    }

    actualizar() {
        if (this.sacudida > 0) this.sacudida--;
        if (this.destello > 0) this.destello--;
    }

    dibujar(ctx, fuente) {
        if (!this.activo) return;

        let ox = 0;
        if (this.sacudida > 0) {
            ox = Math.floor(Math.random() * 7) - 3;
        }

        const rx = this.cx - this.ancho / 2 + ox;
        const ry = this.cy - this.alto / 2;

        let colorFondo = this.sacudida > 0 ? ROJO : AZUL;
        let colorBorde = NEGRO_PURO;

        if (this.destello > 0) {
            if (Math.floor(this.destello / 3) % 2 === 0) {
                colorFondo = VERDE;
                colorBorde = NEGRO_PURO;
            } else {
                colorFondo = [180, 255, 180];
                colorBorde = VERDE;
            }
        }

        ctx.fillStyle = `rgb(${colorFondo.join(',')})`;
        ctx.strokeStyle = `rgb(${colorBorde.join(',')})`;
        ctx.lineWidth = 3;

        this._roundRect(ctx, rx, ry, this.ancho, this.alto, 12);
        ctx.fill();
        ctx.stroke();

        const lineas = this._ajustarTexto(ctx, fuente);
        const altoLinea = 24;
        const totalAlto = lineas.length * altoLinea;
        let sy = this.cy - totalAlto / 2;

        ctx.fillStyle = `rgb(${BLANCO.join(',')})`;
        ctx.font = `bold ${fuente}px Arial`;
        ctx.textAlign = 'center';

        for (const linea of lineas) {
            ctx.fillText(linea, this.cx + ox, sy + fuente * 0.7);
            sy += altoLinea;
        }
    }

    _ajustarTexto(ctx, fuente) {
        const palabras = this.texto.split(' ');
        const lineas = [];
        let linea = '';
        const maxAncho = this.ancho - 14;

        ctx.font = `bold ${fuente}px Arial`;

        for (const p of palabras) {
            const prueba = (linea + ' ' + p).trim();
            const metrics = ctx.measureText(prueba);

            if (metrics.width <= maxAncho) {
                linea = prueba;
            } else {
                if (linea) lineas.push(linea);
                linea = p;
            }
        }
        if (linea) lineas.push(linea);

        return lineas;
    }

    _roundRect(ctx, x, y, w, h, r) {
        ctx.beginPath();
        ctx.moveTo(x + r, y);
        ctx.lineTo(x + w - r, y);
        ctx.quadraticCurveTo(x + w, y, x + w, y + r);
        ctx.lineTo(x + w, y + h - r);
        ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
        ctx.lineTo(x + r, y + h);
        ctx.quadraticCurveTo(x, y + h, x, y + h - r);
        ctx.lineTo(x, y + r);
        ctx.quadraticCurveTo(x, y, x + r, y);
        ctx.closePath();
    }
}